src/catchup/CatchupManagerImpl.cpp

```diff
@@ -93,4 +200,105 @@ CatchupManagerImpl::logAndUpdateCatchupStatus(bool contiguous)
 {
     logAndUpdateCatchupStatus(contiguous, getStatus());
 }
+
+bool
+CatchupManagerImpl::hasBufferedLedger() const
+{
+    return !mSyncingLedgers.empty();
+}
+
+LedgerCloseData const&
+CatchupManagerImpl::getBufferedLedger() const
+{
+    if (!hasBufferedLedger())
+    {
+        throw std::runtime_error(
+            "getBufferedLedger called when mSyncingLedgers is empty!");
+    }
+
+    return mSyncingLedgers.front();
+}
+
+void
+CatchupManagerImpl::popBufferedLedger()
+{
+    if (!hasBufferedLedger())
+    {
+        throw std::runtime_error(
+            "popBufferedLedger called when mSyncingLedgers is empty!");
+    }
+
+    mSyncingLedgers.pop_front();
+}
+
+void
+CatchupManagerImpl::syncMetrics()
+{
+    mSyncingLedgersSize.set_count(mSyncingLedgers.size());
+}
+
+void
+CatchupManagerImpl::trimAndReset()
+{
+    assert(mCatchupWork);
+
+    logAndUpdateCatchupStatus(true);
+    mCatchupWork.reset();
+    trimSyncingLedgers();
+}
+
+void
+CatchupManagerImpl::addToSyncingLedgers(LedgerCloseData const& ledgerData)
+{
+    mSyncingLedgers.push_back(ledgerData);
+
+    CLOG(INFO, "Ledger") << "Close of ledger " << ledgerData.getLedgerSeq()
+                         << " buffered";
+}
+
+void
+CatchupManagerImpl::startOnlineCatchup()
+{
+    assert(mSyncingLedgers.size() > 1);
+
+    // catchup just before first buffered ledger that way we will have a
+    // way to verify history consistency - compare previousLedgerHash of
+    // buffered ledger with last one downloaded from history
+    auto firstBufferedLedgerSeq = mSyncingLedgers.front().getLedgerSeq();
+    auto hash = make_optional<Hash>(
+        mSyncingLedgers.front().getTxSet()->previousLedgerHash());
+    startCatchup({LedgerNumHashPair(firstBufferedLedgerSeq - 1, hash),
+                  getCatchupCount(), CatchupConfiguration::Mode::ONLINE},
+                 nullptr);
+}
+
+void
+CatchupManagerImpl::trimSyncingLedgers()
+{
+    // look for the latest checkpoint
+    auto rit = mSyncingLedgers.rbegin();
+    auto rend = mSyncingLedgers.rend();
+    while (rit != rend)
+    {
+        auto ledger = rit->getLedgerSeq();
+        auto nextCheckpoint =
```

Comment:
Hm, I think this might be wrong? This will be a multiple of 64, i.e. first-ledger-in-a-checkpoint.